### ISSUE TITLE
field 'ctx' has incomplete type

### ISSUE THREAD

**Issue by drjosephliu** [warmcat/libwebsockets#1946]
Getting the same error as in issue #1662 :

```
/vagrant/include/libwebsockets/lws-genhash.h:82:18: error: field ‘ctx’ has incomplete type ‘HMAC_CTX’ {aka ‘hmac_ctx_st’}
   82 |         HMAC_CTX ctx;
```

Running on libwebsockets 3.2.1-1.fc31, libev 4.27-1.fc31 on Fedora 31.

**Comment by drjosephliu** [comment]
Hey so I did try looking for a `./build/CMakeFiles/CMakeError.log` but it isn't there. Is this a linux-specific thing? FWIW I'm using cmake 3.17.3 and make 4.2.1. 

I do have a `./build/CMakeFiles/CMakeOutput.log` file however, but there is no mention of libwebsockets, openssl, or hmac in it.

**Comment by lws-team** [comment]
Hm... I develop lws on Fedora... there's no way to get this problem out of the box with stock pieces and following the recommended build flow.

Similarly, there's no way to run the recommended build flow and not get failed tests, CMakeLists.txt tests for some things Fedora pieces do not have, and consequently creates CMakeError.log.

Is thete something relevant to your actual build flow you have not yet mentioned?

**Comment by OlehKulykov** [comment]
Hi, I have the same issue and both ```build/CMakeFiles/CMakeError.log``` and ```build/CMakeFiles/CMakeOutput.log``` are attached.
Hope this will help.
My env. is:
- OS: MacOS ```10.15.5 (19F101)```
- CMake: ```3.17.3```
- libwebsockets branch: ```v4.0-stable```
- CMake command: ```cmake ..```
- Make command: ```make```
- OpenSSL version: ```1.1.1g```
- Console output with compile error:
```
$ make
Scanning dependencies of target websockets
[  1%] Building C object CMakeFiles/websockets.dir/lib/core/alloc.c.o
In file included from /Users/okulykov/Downloads/libwebsockets/lib/core/alloc.c:25:
In file included from /Users/okulykov/Downloads/libwebsockets/lib/core/private-lib-core.h:135:
In file included from /Users/okulykov/Downloads/libwebsockets/include/libwebsockets.h:600:
/Users/okulykov/Downloads/libwebsockets/include/libwebsockets/lws-genhash.h:80:18: error: field has incomplete
      type 'HMAC_CTX' (aka 'struct hmac_ctx_st')
        HMAC_CTX ctx;
                 ^
/opt/local/include/openssl/ossl_typ.h:102:16: note: forward declaration of 'struct hmac_ctx_st'
typedef struct hmac_ctx_st HMAC_CTX;
               ^
1 error generated.
make[2]: *** [CMakeFiles/websockets.dir/lib/core/alloc.c.o] Error 1
make[1]: *** [CMakeFiles/websockets.dir/all] Error 2
make: *** [all] Error 2
```
[CMakeError.log](https://github.com/warmcat/libwebsockets/files/4807880/CMakeError.log)
[CMakeOutput.log](https://github.com/warmcat/libwebsockets/files/4807881/CMakeOutput.log)

**Comment by lws-team** [comment]
I think these issues are unrelated, but from your output

```
Determining if the function HMAC_CTX_new exists failed with the following output:
Change Dir: /Users/okulykov/Downloads/libwebsockets/build/CMakeFiles/CMakeTmp
...
Linking C executable cmTC_8dcc9
/opt/local/bin/cmake -E cmake_link_script CMakeFiles/cmTC_8dcc9.dir/link.txt --verbose=1
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -Wall -Wsign-compare -Wstrict-aliasing -Wuninitialized -Werror -fvisibility=hidden -Wundef  -Wtype-limits -Wignored-qualifiers   -Wno-deprecated-declarations -pthread -Wno-error=unused-command-line-argument -DCHECK_FUNCTION_EXISTS=HMAC_CTX_new -pthread -Wno-error=unused-command-line-argument -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names   CMakeFiles/cmTC_8dcc9.dir/CheckFunctionExists.c.o  -o cmTC_8dcc9  /opt/local/lib/libssl.dylib /opt/local/lib/libcrypto.dylib -lssl -lcrypto -lm -ldl 
ld: library not found for -lssl
```

Master builds fine in CI on an actual Mac Mini here

```
-- Looking for HMAC_CTX_new
-- Looking for HMAC_CTX_new - found
```

https://libwebsockets.org/sai/index.html?task=46498dc8217fc2a9a9d18a49e25ae94418a103899a1f686c68fba6afc45319a1

What happens if you do the same thing on master?

**Comment by lszl84** [comment]
This still reproduces on MacOS using current master and OpenSSL's master branch.

Is this project compatible with OpenSSL 1.1.x?

**Comment by lws-team** [comment]
> This still reproduces on MacOS using current master and OpenSSL's master branch.

This is to do with what args you personally are or are not giving cmake.

> Is this project compatible with OpenSSL 1.1.x?

Yes, and it's built with that in CI for both mac mini x86_64

https://libwebsockets.org/sai/index.html?task=999c8f33441c55728d519ce531c775693989f1d5242a04e003524edb2296d816

and M1

https://libwebsockets.org/sai/index.html?task=999c8f33441c55728d519ce531c77569d11aba63cf00fcb4a4add35eacf66bf2

Look in `build/CMakeFiles/CMakeError.log` and find out what your actual problem is... refer to how the CI configures cmake as a reference

```
cmake .. -DLWS_WITH_SUL_DEBUGGING=1 -DCMAKE_SYSTEM_PREFIX_PATH=/opt/homebrew -DLWS_OPENSSL_INCLUDE_DIRS=/opt/homebrew/Cellar/openssl@1.1/1.1.1h/include '-DLWS_OPENSSL_LIBRARIES=/opt/homebrew/Cellar/openssl@1.1/1.1.1h/lib/libssl.dylib;/opt/homebrew/Cellar/openssl@1.1/1.1.1h/lib/libcrypto.dylib' -DLWS_WITH_MINIMAL_EXAMPLES=1 -DLWS_WITH_MINIMAL_EXAMPLES=1
```

**Comment by lszl84** [comment]
Thanks, I think I found the issue.

I was confused by the find_package's suggestion:

```
CMake Error at /usr/local/Cellar/cmake/3.19.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:218 (message):
  Could NOT find OpenSSL, try to set the path to OpenSSL root folder in the
  system variable OPENSSL_ROOT_DIR (missing: OPENSSL_INCLUDE_DIR)
```

It says "system variable OPENSSL_ROOT_DIR", so I tried:

```
OPENSSL_ROOT_DIR=/usr/local/Cellar/openssl@1.1/1.1.1i cmake -S. -Bbuild && cmake --build build
```

The find_package seemed to find OpenSSL, but somehow used Apple's default libraries /usr/lib/libcrypto.dylib, instead of Homebrew's one in provided OPENSSL_ROOT_DIR. That caused an error when looking for HMAC_CTX_new:

```
ld: cannot link directly with dylib/framework, your binary is not an allowed client of /usr/lib/libcrypto.dylib for architecture x86_64
```

So this seems to be CMake's bug (https://stackoverflow.com/questions/58446253/xcode-11-ld-error-your-binary-is-not-an-allowed-client-of-usr-lib-libcrypto-dy#comment103570984_58446253). 

**TL;DR:** In case anyone experiences this problem on MacOS, make sure to pass your Homebrew's OpenSSL install dir as a CMake variable, not a system variable. Use for example:

```
cmake -S. -Bbuild -DOPENSSL_ROOT_DIR=/usr/local/Cellar/openssl@1.1/1.1.1i && cmake --build build
```

**Comment by rajhlinux** [comment]
Hello, I have the same error as OP:

```
root@FreeBSD:/usr/home/user/Downloads/ftx-master/cpp/bin # make
[ 18%] Built target rest
Consolidate compiler generated dependencies of target ws
[ 27%] Building CXX object src/ws/CMakeFiles/ws.dir/client.cc.o
In file included from /usr/home/user/Downloads/ftx-master/cpp/src/ws/client.cc:2:
/usr/home/user/Downloads/ftx-master/cpp/src/util/Encoding.h:11:14: fatal error: field has incomplete type 'HMAC_CTX' (aka 'hmac_ctx_st')
    HMAC_CTX ctx;
             ^
/usr/local/include/openssl/types.h:132:16: note: forward declaration of 'hmac_ctx_st'
typedef struct hmac_ctx_st HMAC_CTX;
               ^
1 error generated.
*** Error code 1

Stop.
make[2]: stopped in /usr/home/user/Downloads/ftx-master/cpp/bin
*** Error code 1

Stop.
make[1]: stopped in /usr/home/user/Downloads/ftx-master/cpp/bin
*** Error code 1

Stop.
make: stopped in /usr/home/user/Downloads/ftx-master/cpp/bin
root@FreeBSD:/usr/home/user/Downloads/ftx-master/cpp/bin #
```

Thanks.

**Comment by lws-team** [comment]
Read the second section here

https://libwebsockets.org/git/libwebsockets/tree/READMEs/README.cmake.md
